extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+         .emit();
+      throw Throwables.propagate(e);
+    }
+
+    started = true;
+    log.info("Started KafkaSupervisor[%s].", dataSource);
+  }
+
+  @Override
+  public void stop()
+  {
+    Preconditions.checkState(started, "not started");
+
+    try {
+      exec.shutdownNow();
+      consumer.close();
+      started = false;
+
+      log.info("KafkaSupervisor[%s] has stopped.", dataSource);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception stopping KafkaSupervisor[%s]", dataSource)
+         .emit();
+    }
+  }
+
+  @Override
+  public Optional<SupervisorSpec> getSpec()
+  {
+    return Optional.fromNullable((SupervisorSpec) this.spec);
+  }
+
+  private void mainLoop()
+  {
+    try {
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          killUnownedKafkaTasks();
+          updatePartitionDataFromKafka();
+          updateTaskMetadata();
+          checkTaskState();
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Exception in KafkaSupervisor[%s]", dataSource)
+             .emit();
+        }
+
+        Thread.sleep(config.getRunPeriod());
+      }
+    }
+    catch (InterruptedException e) {
+      log.info("KafkaSupervisor[%s] interrupted, exiting.", dataSource);
+    }
+  }
+
+  private static String getRandomId()
+  {
+    final StringBuilder suffix = new StringBuilder(8);
+    for (int i = 0; i < Ints.BYTES * 2; ++i) {
+      suffix.append((char) ('a' + ((RANDOM.nextInt() >>> (i * 4)) & 0x0F)));
+    }
+    return suffix.toString();
+  }
+
+  private KafkaConsumer<byte[], byte[]> getKafkaConsumer(String bootstrapServers)
+  {
+    final Properties props = new Properties();
+    props.putAll(spec.getIoConfig().getConsumerProperties());
+
+    props.setProperty("enable.auto.commit", "false");
+    props.setProperty("metadata.max.age.ms", "10000");
+    props.setProperty("bootstrap.servers", bootstrapServers);
+    props.setProperty("group.id", String.format("kafka-supervisor-%s", supervisorRunId));
+
+    ClassLoader currCtxCl = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+      return new KafkaConsumer<>(props, new ByteArrayDeserializer(), new ByteArrayDeserializer());
+    }
+    finally {
+      Thread.currentThread().setContextClassLoader(currCtxCl);
+    }
+  }
+
+  private void updatePartitionDataFromKafka()
+  {
+    try {
+      consumer.listTopics(); // updates the consumer's list of partitions from the brokers
+    }
+    catch (Exception e) { // calls to the consumer throw NPEs when the broker doesn't respond
+      log.warn(
+          e,
+          "Unable to get partition data from Kafka for brokers [%s], are the brokers up?",
+          spec.getIoConfig().getKafkaBrokers()
+      );
+      return;
+    }
+
+    List<PartitionInfo> partitions = consumer.partitionsFor(spec.getIoConfig().getTopic());
```

Comment:
I was finding that partitionsFor() didn't actually detect new partitions that were made, even when I had metadata.max.age.ms set, it was still returning cached data. I just confirmed it now by looking at the implementation. Calling listTopics() results in a call to the server to update the metadata. You're right, the partitionsFor() call is unnecessary.